### PR TITLE
Add support for CRD `ConversionReview` types

### PR DESCRIPTION
**Pull request by MikailBag**
Signed-off-by: Mikail Bagishov <bagishov.mikail@yandex.ru>

## Motivation

Part of #992.

## Solution

Compared to version in the old PR, I renamed `low_level.rs -> types.rs`, renamed module `conversion.rs -> mod/conversion.rs` to align with rest of the code base, and made submodule private.

### CODE REVIEW

File: kube-core/src/conversion/types.rs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    /// if conversion failed.
+    #[serde(rename = "convertedObjects")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub converted_objects: Option<Vec<serde_json::Value>>,
+}
+
+impl ConversionResponse {
+    /// Creates a new response, matching provided request.
+    /// Returned response is empty: you should call `success` or `error`
+    /// to get a complete `ConversionResponse`.
```
Comment (clux) [inline kube-core/src/conversion/types.rs:95]:
```suggestion
    /// Creates a blank response, matching provided request.
    ///
    /// This response must be finalized with one of:
    /// - [`ConversionResponse::ok`] when conversion succeeded
    /// - [`ConversionResponse::error`] when conversion failed
```

File: kube-core/src/conversion/types.rs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+                reason: String::new(),
+                details: None,
+            },
+            converted_objects: None,
+        }
+    }
+
+    /// Creates successful conversion response.
+    /// `converted_objects` must specify objects in the exact same order as on input.
+    pub fn success(mut self, converted_objects: Vec<serde_json::Value>) -> Self {
```
Comment (clux) [inline kube-core/src/conversion/types.rs:113]: given we have `error` as a method name the more appropriate counterpart should be `ok`

File: kube-core/src/conversion/types.rs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    /// `message` and `reason` will be returned to the apiserver.
+    pub fn error(mut self, message: &str, reason: &str) -> Self {
+        self.result = Status::failure(message, reason);
+        self
+    }
+
+    /// Creates failed conversion response, not matched with any request.
+    /// You should only call this function when request couldn't be parsed into [`ConversionRequest`].
+    /// Otherwise use `error`.
+    pub fn unmatched_error(message: &str, reason: &str) -> Self {
```
Comment (clux) [inline kube-core/src/conversion/types.rs:130]: this can be called `invalid` in line with admission

File: kube-core/src/conversion/types.rs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+#[cfg(test)]
+mod tests {
+    use super::ConversionReview;
+
+    #[test]
+    fn simple_request_parses() {
+        // this file contains dump of real request generated by kubernetes v1.22
+        let data = include_str!("./test_data/simple.json");
+        // check that we can parse this request
+        let _: ConversionReview = serde_json::from_str(data).unwrap();
```
Comment (clux) [inline kube-core/src/conversion/types.rs:176]: can we test the methods here and turn it into a Request and then a Response here as a minimum?

File: kube-core/src/response.rs [hunk 5]
```diff
@@ -1,20 +1,20 @@
 //! Generic api response types
-use serde::Deserialize;
+use serde::{Deserialize, Serialize};
 
 /// A Kubernetes status object
 ///
 /// Equivalent to Status in k8s-openapi except we have have simplified options
```
Comment (clux) [inline kube-core/src/response.rs:6]: I would remove this line while we are here, there are more ergonomic differences now

File: kube-core/src/conversion/types.rs (hunk 2 above)
Comment (clux) [reply kube-core/src/conversion/types.rs:113]: alternatively, we can rename `::error` to `::failure`

File: kube-core/src/response.rs [hunk 6]
```diff
@@ -35,8 +35,53 @@ pub struct Status {
...
+        self.code = code;
+        self
+    }
+
+    /// Adds details to the `Status`
+    pub fn with_details(mut self, details: StatusDetails) -> Self {
+        self.details = Some(details);
+        self
+    }
+}
```
Comment (clux) [inline kube-core/src/response.rs:72]: can we also add:

```
pub fn is_failure(&self) -> bool;
pub fn is_success(&self) -> bool;
```

while we are here? :-)

**Review by clux — COMMENTED** [review]
minor comments on documentation and naming, but this part is starting to look good

File: kube-core/src/conversion/types.rs [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    pub response: Option<ConversionResponse>,
+}
+
+/// Part of ConversionReview which is set on input (i.e. generated by apiserver)
+#[derive(Serialize, Deserialize)]
+pub struct ConversionRequest {
+    /// Copied from the corresponding consructing [`ConversionReview`].
+    /// This field is not part of the Kubernetes API, it's consumed only by `kube`.
+    #[serde(skip)]
+    pub types: Option<TypeMeta>,
```
[the remark this answers is not in the document]
Comment (MikailBag) [reply kube-core/src/conversion/types.rs:36]: No strong reason. It simplifies manual creation of `ConversionRequest` a little bit (i.e. it may be surprising for user to see non-optional field which is not set by apiserver).

File: kube-core/src/conversion/types.rs [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    /// Outcome of the conversion operation.
+    /// Success: all objects were successfully converted
+    /// Failure: at least one object could not be converted.
+    /// It is recommended that conversion fails as rare as possible.
+    pub result: Status,
+    /// Converted objects in the same order as in the request. Should be empty
+    /// if conversion failed.
+    #[serde(rename = "convertedObjects")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub converted_objects: Option<Vec<serde_json::Value>>,
```
[the remark this answers is not in the document]
Comment (MikailBag) [reply kube-core/src/conversion/types.rs:89]: As far as I remember, without this Option example webhook failed on [this](https://github.com/kube-rs/kube-rs/pull/999/files#diff-3ba17590977dafaa6f5ba1ff1d3db64d0531ebfaf3bb23221edb2978fe756273R50) part of input.

File: kube-core/src/conversion/types.rs [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    pub types: Option<TypeMeta>,
+    /// Random uid uniquely identifying this conversion call
+    pub uid: String,
+    /// The API group and version the objects should be converted to
+    #[serde(rename = "desiredAPIVersion")]
+    pub desired_api_version: String,
+    /// The list of objects to convert.
+    /// May contain one or more objects, in one or more versions.
+    // This field used raw Value instead of Object/DynamicObject to simplify
+    // further downcasting.
```
[the remark this answers is not in the document]
Comment (MikailBag) [reply kube-core/src/conversion/types.rs]: Probably done

File: kube-core/src/conversion/types.rs [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    pub objects: Vec<serde_json::Value>,
+}
+
+impl ConversionRequest {
+    /// Extracts request from the [`ConversionReview`].
+    /// Fails if review has missing request.
```
[the remark this answers is not in the document]
Comment (MikailBag) [reply kube-core/src/conversion/types.rs]: Done

File: kube-core/src/response.rs (hunk 5 above)
Comment (MikailBag) [reply kube-core/src/response.rs]: Done

File: kube-core/src/response.rs (hunk 6 above)
Comment (MikailBag) [reply kube-core/src/response.rs:86]: Done

File: kube-core/src/conversion/types.rs (hunk 2 above)
Comment (MikailBag) [reply kube-core/src/conversion/types.rs:122]: I'll stick with the second way (rename error to failure), because it better aligns with the `Success` and `Failure` naming of statuses.

File: kube-core/src/conversion/types.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (clux) [reply kube-core/src/conversion/types.rs:96]: If that is a legit response that Kubernetes accepts then that's OK, but it does seem like the docs indicate we should be putting in an explicit empty vec in the failure case rather than a null.

File: kube-core/src/conversion/types.rs [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    /// This field is not part of the Kubernetes API, it's consumed only by `kube`.
+    #[serde(skip)]
+    pub types: Option<TypeMeta>,
+    /// Copy of .request.uid
+    pub uid: String,
+    /// Outcome of the conversion operation.
+    /// Success: all objects were successfully converted
+    /// Failure: at least one object could not be converted.
+    /// It is recommended that conversion fails as rare as possible.
+    pub result: Status,
```
[the remark this answers is not in the document]
Comment (MikailBag) [reply kube-core/src/conversion/types.rs:89]: Done

File: kube-core/src/conversion/types.rs (hunk 4 above)
Comment (MikailBag) [reply kube-core/src/conversion/types.rs]: Done

File: kube-core/src/conversion/types.rs (hunk 3 above)
Comment (MikailBag) [reply kube-core/src/conversion/types.rs]: Done

File: kube-core/src/conversion/types.rs (hunk 2 above)
Comment (MikailBag) [reply kube-core/src/conversion/types.rs:123]: Done

File: kube-core/src/conversion/types.rs (hunk 1 above)
Comment (MikailBag) [reply kube-core/src/conversion/types.rs]: Done

File: kube-core/src/conversion/types.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (MikailBag) [reply kube-core/src/conversion/types.rs]: I changed implementation so that field is typed as `Vec<Value>` (and null is special-cased during parsing).

File: kube-core/src/conversion/types.rs [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+    /// Extracts request from the [`ConversionReview`].
+    /// Fails if review has missing request.
+    pub fn from_review(review: ConversionReview) -> Result<Self, ConvertConversionReviewError> {
+        match review.request {
+            Some(mut req) => {
+                req.types = Some(review.types);
+                Ok(req)
+            }
+            None => Err(ConvertConversionReviewError),
+        }
```
[the remark this answers is not in the document]
Comment (MikailBag) [reply kube-core/src/conversion/types.rs]: Done

File: kube-core/src/conversion/types.rs [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+use thiserror::Error;
+
+/// The `kind` field in [`TypeMeta`]
+pub const META_KIND: &str = "ConversionReview";
+/// The `api_version` field in [`TypeMeta`] on the v1 version
+pub const META_API_VERSION_V1: &str = "apiextensions.k8s.io/v1";
+
+#[derive(Debug, Error)]
+#[error("request missing in ConversionReview")]
+/// Returned when `AdmissionReview` cannot be converted into `AdmissionRequest`
```
Comment (clux) [inline kube-core/src/conversion/types.rs:12]: copy-pasted docs are slightly wrong

File: kube-core/src/conversion/types.rs [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+        self.result = Status::success();
+        self.converted_objects = converted_objects;
+        self
+    }
+
+    /// Creates failed conversion response (discouraged)
+    ///
+    /// `request_uid` must be equal to the `.uid` field in the request.
+    /// `message` and `reason` will be returned to the apiserver.
+    pub fn failure(mut self, message: &str, reason: &str) -> Self {
```
Comment (clux) [inline kube-core/src/conversion/types.rs:141]: i think this should take a `Status` as a param rather than `message` and `reason` (two &str arguments)

Comment (clux) [reply kube-core/src/conversion/types.rs:141]: ditto for invalid below

File: kube-core/src/response.rs [hunk 15]
```diff
@@ -35,8 +33,69 @@ pub struct Status {
...
+            status: Some(StatusSummary::Success),
+            code: 0,
+            message: String::new(),
+            reason: String::new(),
+            details: None,
+        }
+    }
+
+    /// Returns an unsuccessful `Status`
+    pub fn failure(message: &str, reason: &str) -> Self {
```
Comment (clux) [inline kube-core/src/response.rs:49]: maybe we should split this up into a `failure` with `message` as the only param and have a `with_reason(reason)` function on `mut self` (or other way around if reason is the most important one - don't remember right now)

File: kube-core/src/conversion/types.rs [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+fn parse_converted_objects<'de, D>(de: D) -> Result<Vec<serde_json::Value>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    #[derive(Deserialize)]
+    #[serde(untagged)]
+    enum Helper {
+        List(Vec<serde_json::Value>),
+        Null(()),
+    }
```
Comment (clux) [inline kube-core/src/conversion/types.rs:108]: so this is basically accounting for an omitted `converted_objects` or `converted_objects: null` and the normal `converted_objects:[...]`? could this not be handled with a `#[serde(default)]` instead?

File: kube-core/src/response.rs (hunk 15 above)
Comment (clux) [reply kube-core/src/response.rs:49]: actually, these two are very linked, it is probably a mistake to let people do one and not the other. ignore this.

while i am a bit uncomfortable with the double &str argument (which can be swapped around accidentally), there are probably smarter stuff we can do down the line to convert an `Error` into a `Status` by taking advantage of the `Display` and `Debug` traits for the error to fill these in a less error prone way.

File: kube-core/src/conversion/types.rs (hunk 13 above)
Comment (MikailBag) [reply kube-core/src/conversion/types.rs]: Fixed

File: kube-core/src/conversion/types.rs (hunk 16 above)
Comment (MikailBag) [reply kube-core/src/conversion/types.rs:108]:
```
thread 'conversion::types::tests::simple_request_parses' panicked at 'called `Result::unwrap()` on an `Err` value: Error("invalid type: null, expected a sequence", line: 50, column: 32)', kube-core/src/conversion/types.rs:207:49
```
AFAIU, `#[serde(default)]` applies when value is completely missing, but apiserver (at least the one I've tested my changes on) sets `null`.

File: kube-core/src/conversion/types.rs (hunk 14 above)
Comment (MikailBag) [reply kube-core/src/conversion/types.rs]: Done

File: kube-core/src/response.rs (hunk 15 above)
Comment (MikailBag) [reply kube-core/src/response.rs:49]: I can think of
```
fn from_error<E: Display>(err: &E, reason: &str) -> Self;
```
which uses e.to_string() as message (I don't see a way to infer reason from error, especially given guidelines that it shoud be short and PascalCase).

File: kube-core/src/conversion/types.rs (hunk 16 above)
Comment (clux) [reply kube-core/src/conversion/types.rs:108]: Well, then we are back to the same question of whether or not `converted_objects` can be `null`.
**If** your `simple.json` is actually retrieved from the apiserver, then I would rather have `Option<Vec<..>>` because then we can use the simpler `#[serde(default)]`.
...but, neither the [source](https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/v1/types.go#L474-L478), nor the [protobufs](https://github.com/kube-rs/k8s-pb/blob/d60271637aac5bb854d529f99d5e6a23154acd06/k8s-pb/src/apiextensions_apiserver/pkg/apis/apiextensions/v1/mod.rs#L29) list it as optional, so i am curious whether your test case is actually real. if it is not, i would not bother trying to handle it, if it is, then it's probably an upstream Kubernetes bug.

File: kube-core/src/response.rs (hunk 15 above)
Comment (clux) [reply kube-core/src/response.rs:49]: Users could possibly do it in their app, if they did a convention like that if the Display was always PascalCase, and then they can take the Debug as the reason, but users would have to stay on top of their errors if so. We couldn't enforce that though, so I'd rather leave that blank and experiment in controller-rs with something like that.

File: kube-core/src/conversion/types.rs (hunk 16 above)
Comment (MikailBag) [reply kube-core/src/conversion/types.rs:108]: > If your simple.json is actually retrieved from the apiserver

Yes. I've written a web server which logs incoming requests (without deserializing to Rust structs), and specified it as a conversion webhook.

I *guess* [here](https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apiserver/conversion/webhook_converter.go#L156) a `nil` slice is assigned, which later gets converted into `null`.

I guess I should try the same on a vanilla `kind` cluster and see what happens (because I used a managed k8s service, hence it is *possible* its admission code was patched).

File: kube-core/src/admission.rs [hunk 17]
```diff
@@ -307,10 +305,7 @@ impl AdmissionResponse {
             },
             uid: Default::default(),
             allowed: false,
-            result: Status {
-                reason: Some(reason.to_string()),
-                ..Default::default()
-            },
+            result: Status::failure("", &reason.to_string()),
```
Comment (clux) [inline kube-core/src/admission.rs:308]: Minor nit (and technically unrelated to the PR), but it would make more sense to have `message` be the field passed by `invalid`, and we can set the shorter `reason` to be `InvalidRequest`.

File: kube-core/src/conversion/types.rs (hunk 16 above)
Comment (clux) [reply kube-core/src/conversion/types.rs:108]: Well, the source hasn't changed in 3 years so if you are seeing it, we will likely see it from others. So, it probably makes sensi to keep this more defensive deserializer in here. It's not a big deal, and it even makes the user expectation match the go structs.

**Comment by clux** [comment]
Only one tiny nit left on the admission code that was cleaned up, but am otherwise happy with merging this. Thanks for all the back and forth on this.

File: kube-core/src/conversion/types.rs (hunk 16 above)
Comment (MikailBag) [reply kube-core/src/conversion/types.rs:108]: I confirmed this behavior using kind v0.15.0 and vanilla k8s v1.25.0.
I'm marking this conversation as resolved: while hacks on our side are unfortunate, we can't avoid them right now.

**Comment by MikailBag** [comment]
CI is red, but this looks like issue with k3d action

File: kube-core/src/admission.rs (hunk 17 above)
Comment (MikailBag) [reply kube-core/src/admission.rs]: Done

**Review by clux — APPROVED** [review]

**Comment by clux** [comment]
> CI is red, but this looks like issue with k3d action

Yeah, had the same issue on the k8s-openapi bump PR, so probably not resolvable here without pinning k3d. I'll give it one last re-run as a check before merging.

**Comment by clux** [comment]
Yeah, consistently k3d action failing. I'll fix that in master with some pinning if necessary. Going to deal with this anyway in #1008.

Going to merge this. Thanks a lot for all the work on this!
